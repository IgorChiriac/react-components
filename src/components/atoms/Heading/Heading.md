### Summary

1. Whenever you want to render a heading on any UI at Zopa, use the `<Heading />` component
2. Supply an `as` prop to specify which level to render: "h1" | "h2" | "h3" | "h4"
3. Depending on the rendered tag through `as` the headings will render at bigger or smaller size
4. Its size can be controlled through the `size` prop as well for more granular control
5. Its weight can't be customised
6. It comes with bottom white-space (margin) by default but you can disabled that through the `mb` prop

### Examples

- Variations

```tsx
import { Fragment } from 'react';
import { Heading } from '@zopauk/react-components';

<Fragment>
  <Heading as="h1">Heading level 1</Heading>
  <Heading as="h2">Heading level 2</Heading>
  <Heading as="h3">Heading level 3</Heading>
  <Heading as="h4" mb={false}>
    Heading level 4
  </Heading>
</Fragment>;
```

- With white-space

```tsx
import { Fragment } from 'react';
import { Heading } from '@zopauk/react-components';

<Fragment>
  <Heading as="h3">With bottom white-space</Heading>
  <Heading as="h3" mb={false}>
    Without bottom white-space
  </Heading>
  <Heading as="h3" mb={false}>
    Without bottom white-space
  </Heading>
</Fragment>;
```

- Sizing

```jsx
import { Fragment } from 'react';
import { Heading } from '@zopauk/react-components';

<Fragment>
  <Heading as="h2">As a h2 I render at 28px by default</Heading>
  <Heading as="h4">As a h4 I render at 20px by default</Heading>
  <Heading as="h2" size="h4">
    I render with a h2 tag but small
  </Heading>
  <Heading as="span" size="h3">
    I render as a {'<span>'} and big
  </Heading>
</Fragment>;
```

- Colors

```tsx { "props": { "style": { "backgroundColor": "rgb(244, 248, 246)", "border": "none" } } }
import { Fragment } from 'react';
import { Heading, colors } from '@zopauk/react-components';

<Fragment>
  <Heading as="h4" color={colors.neutral.white}>
    White
  </Heading>
  <Heading as="h4" color={colors.neutral.nearDark}>
    Near Dark
  </Heading>
  <Heading as="h4" color={colors.neutral.dark} mb={false}>
    Dark
  </Heading>
</Fragment>;
```
