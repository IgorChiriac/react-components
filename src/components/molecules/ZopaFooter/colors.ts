export const neutral900 = '#1C2139';
export const neutral200 = '#A9ACBA';
export const neutral400 = '#6F748B';
export const neutral800 = '#282F52';
