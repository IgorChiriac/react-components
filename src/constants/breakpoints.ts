export const breakpoints = {
  desktop: 1280,
  phone: 600,
  tablet: 720,
} as const;

export type TDeviceSizes = keyof typeof breakpoints;
