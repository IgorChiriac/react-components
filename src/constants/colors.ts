export type TTextHexColors = '#FFFFFF' | '#C5C6CD' | '#0D0A38' | '#1EC06A' | '#EE0505';
export type THeadingHexColors = '#FFFFFF' | '#0D0A38';
export type TLinkHexColors = '#FFFFFF' | '#0D0A38';

export const colors = {
  base: {
    primary: '#00B9A7',
    secondary: '#4A3EDE',
  },
  neutral: {
    white: '#FFFFFF',
    light: '#EAEAEE',
    medium: '#C5C6CD',
    dark: '#0D0A38',
  },
  semantic: {
    success: '#1EC06A',
    alert: '#FFB428',
    error: '#EE0505',
  },
};
