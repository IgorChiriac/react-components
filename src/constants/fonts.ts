export const openSans: string =
  '"Open Sans", -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol"';
export const alverata: string = 'Alverata, "Arial Black", Verdana, sans-serif';
