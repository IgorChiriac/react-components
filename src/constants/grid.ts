const grid = {
  breakpoints: {
    l: 992,
    m: 768,
    s: 576,
    xl: 1200,
    xs: 0,
  },
  cols: 12,
  gutter: 16,
  width: {
    l: 960,
    m: 720,
    s: 540,
    xl: 1140,
  },
};

export default grid;
