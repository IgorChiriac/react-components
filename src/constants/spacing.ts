export const spacing = [0, 4, 8, 16, 24, 32, 40, 56, 64, 104];
