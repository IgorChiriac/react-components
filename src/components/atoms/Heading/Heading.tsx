import React, { FC, HTMLAttributes } from 'react';
import styled from 'styled-components';
import Text from '../Text/Text';
import { typography } from '../../../constants/typography';

const {
  sizes: { heading: headingSizes },
} = typography;

interface IStyledHeadingProps extends HTMLAttributes<HTMLHeadingElement> {
  /**
   * The HTML5 tag you want to render your heading, it's used to determine the size of the heading as well.
   */
  as: 'h1' | 'h2' | 'h3' | 'h4';
  /**
   * Whether to add some margin below the rendered heading or not. Applied by default.
   * @default false
   */
  mb?: boolean;
}

const StyledHeading = styled(Text)<IStyledHeadingProps>`
  font-size: ${({ as }) => headingSizes[as]};
  font-family: ${typography.primary};
  font-weight: ${typography.weights.bold};
  line-height: ${typography.lineHeights.heading};
  letter-spacing: -0.5px;
  margin: 0;

  ${({ mb = true }) => mb && 'margin-bottom: 24px'};
`;

const Heading: FC<IStyledHeadingProps> = props => <StyledHeading {...props} />;

export default Heading;
