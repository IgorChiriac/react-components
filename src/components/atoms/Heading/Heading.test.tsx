import { axe } from 'jest-axe';
import React from 'react';
import { render } from '@testing-library/react';
import { colors } from '../../../constants/colors';
import Heading from './Heading';

describe('<Heading />', () => {
  it('renders without  a11y violations', async () => {
    const { container } = render(<Heading as="h1">Text</Heading>);
    const results = await axe(container.innerHTML);

    expect(results).toHaveNoViolations();
    expect(container).toMatchSnapshot();
  });

  it.each`
    tag
    ${'h1'}
    ${'h2'}
    ${'h3'}
    ${'h4'}
  `('it can render with a different HTML tag: $tag', ({ tag }: { tag: 'h1' | 'h2' | 'h3' | 'h4' }) => {
    const { container } = render(<Heading as={tag}>Header</Heading>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it.each`
    hex                     | name
    ${colors.neutral.white} | ${'White'}
    ${colors.neutral.dark}  | ${'Dark'}
  `('can render in different colors: $name â€“ $hex', ({ hex }) => {
    const { container } = render(
      <Heading as="h2" color={hex}>
        Text
      </Heading>,
    );
    expect(container.firstChild).toHaveStyleRule('color', hex);
  });

  it('can render without margin bottom', () => {
    const { container } = render(
      <Heading as="h1" mb={false}>
        Header
      </Heading>,
    );

    expect(container.firstChild).toHaveStyleRule('margin', '0');
  });
});
