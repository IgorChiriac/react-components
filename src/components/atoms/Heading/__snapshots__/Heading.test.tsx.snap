// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Heading /> it can render with a different HTML tag: "h1" 1`] = `
.c0 {
  font-size: 48px;
  font-family: "Open Sans",-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,Helvetica,Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol";
  font-weight: 700;
  line-height: 1.2;
  -webkit-letter-spacing: -0.5px;
  -moz-letter-spacing: -0.5px;
  -ms-letter-spacing: -0.5px;
  letter-spacing: -0.5px;
  margin: 0;
  margin-bottom: 24px;
}

<h1
  class="c0"
>
  Text
</h1>
`;

exports[`<Heading /> it can render with a different HTML tag: "h2" 1`] = `
.c0 {
  font-size: 28px;
  font-family: "Open Sans",-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,Helvetica,Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol";
  font-weight: 700;
  line-height: 1.2;
  -webkit-letter-spacing: -0.5px;
  -moz-letter-spacing: -0.5px;
  -ms-letter-spacing: -0.5px;
  letter-spacing: -0.5px;
  margin: 0;
  margin-bottom: 24px;
}

<h2
  class="c0"
>
  Text
</h2>
`;

exports[`<Heading /> it can render with a different HTML tag: "h3" 1`] = `
.c0 {
  font-size: 24px;
  font-family: "Open Sans",-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,Helvetica,Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol";
  font-weight: 700;
  line-height: 1.2;
  -webkit-letter-spacing: -0.5px;
  -moz-letter-spacing: -0.5px;
  -ms-letter-spacing: -0.5px;
  letter-spacing: -0.5px;
  margin: 0;
  margin-bottom: 24px;
}

<h3
  class="c0"
>
  Text
</h3>
`;

exports[`<Heading /> it can render with a different HTML tag: "h4" 1`] = `
.c0 {
  font-size: 20px;
  font-family: "Open Sans",-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,Helvetica,Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol";
  font-weight: 700;
  line-height: 1.2;
  -webkit-letter-spacing: -0.5px;
  -moz-letter-spacing: -0.5px;
  -ms-letter-spacing: -0.5px;
  letter-spacing: -0.5px;
  margin: 0;
  margin-bottom: 24px;
}

<h4
  class="c0"
>
  Text
</h4>
`;

exports[`<Heading /> renders without  a11y violations 1`] = `
.c0 {
  font-size: 48px;
  font-family: "Open Sans",-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,Helvetica,Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol";
  font-weight: 700;
  line-height: 1.2;
  -webkit-letter-spacing: -0.5px;
  -moz-letter-spacing: -0.5px;
  -ms-letter-spacing: -0.5px;
  letter-spacing: -0.5px;
  margin: 0;
  margin-bottom: 24px;
}

<div>
  <h1
    class="c0"
  >
    Text
  </h1>
</div>
`;
